### DISCUSSION TITLE
Generate file for each type for server development

### DISCUSSION THREAD

**Discussion by Fabyao** [dotansimha/graphql-code-generator#4922] [General] [answered]
Thanks for the great tool! Looking through the documentation I am not sure if its possible to generate a file per type? Given the following schema:

```
type Author {
  id: Int!
  firstName: String!
  lastName: String!
  posts(findTitle: String): [Post]
}

type Post {
  id: Int!
  title: String!
  author: Author!
}

```
Could we generate 2 files:
```
author.ts
post.ts
```

We have a large team and generating all the types in one file will cause merge conflicts.

**Comment by dotansimha** [comment, marked as answer]
Hi! Currently, we don't support that. This could be done with a custom preset for codegen - you can break your schema to smaller pieces and feed it back to the `typescript` plugin.
Please note that we try to avoid these kind of solutions intentionally, since maintaining multiple files is far more complex, and it might cause issues like circular imports and dependencies.

**Reply by JaffParker** [reply]
What kind of issues are we looking at in the case of circular dependencies? I use type-graphql to define my schema in the backend and there's circular dependency galore. JS module system is very well equipped for that.

Also, would you mind pointing me in the direction of docs for creating custom presets? I only found docs for plugins. Or should I just go ahead and "reverse engineer" an existing preset to learn how to make one?

**Reply by dotansimha** [reply]
It's not highly documented. But think of a codegen `preset` as way to dynamically create the `generates` section. For example, with `near-operation-file` preset, we just generate multiple `generates` records based on the loaded operations. The API is super simple. 
Here's a piece of code that might help: https://github.com/dotansimha/graphql-code-generator/blob/master/packages/presets/near-operation-file/src/index.ts#L128-L135

Regarding circular dependencies - with TS types it's simple, also with functions. If someone will mix with that a plugin that generates code, it might be tricky (https://nodejs.org/api/modules.html#modules_cycles).
